designer/client/components/componentCreate/ComponentCreate.tsx

```diff
@@ -1,145 +1,131 @@
-import React, { Component, ChangeEvent, MouseEvent } from "react";
-import { clone, ComponentDef } from "@xgovformbuilder/model";
+import React, {
+  useEffect,
+  useContext,
+  useState,
+  useLayoutEffect,
+  FormEvent,
+} from "react";
+import { ComponentDef } from "@xgovformbuilder/model";
 
 import { i18n } from "../../i18n";
 import { ErrorSummary } from "../../error-summary";
 import { hasValidationErrors } from "../../validations";
-import ComponentTypeEdit from "../../component-type-edit";
+import ComponentTypeEdit from "../../ComponentTypeEdit";
 import { ComponentCreateList } from "./ComponentCreateList";
 import { BackLink } from "../backLink";
 
-import "./ComponentCreate.scss";
-
-type Props = {
-  data: any;
-  onCreate: any;
-  page: any;
-};
-
-type State = {
-  id: string;
-  isSaving: boolean;
-  component?: { type: string; name: string };
-  errors: any;
-};
-
-export class ComponentCreate extends Component<Props, State> {
-  typeEditRef = React.createRef<ComponentTypeEdit>();
-
-  state = {
-    id: "",
-    errors: {},
-    component: undefined,
-    isSaving: false,
-  };
-
-  componentDidMount = async () => {
-    const { data } = this.props;
-    const id = await data.getId();
-    this.setState({ id });
-  };
-
-  onSubmit = async (event: ChangeEvent<HTMLFormElement>) => {
-    event.preventDefault();
-
-    if (this.state.isSaving) return false;
-
-    let validationErrors = this.validate();
-    if (hasValidationErrors(validationErrors)) return false;
-
-    this.setState({ isSaving: true });
+import { Actions } from "../../reducers/component/types";
+import { DataContext } from "../../context";
+import { ComponentContext } from "../../reducers/component/componentReducer";
 
-    const { page, data } = this.props;
-    const { component } = this.state;
-    const copy = clone(data);
+import "./ComponentCreate.scss";
 
-    const updated = copy.addComponent(page.path, component);
-    const saved = await data.save(updated);
-    this.props.onCreate({ data: saved });
+function useComponentCreate(props) {
+  const { data, save } = useContext(DataContext);
+  const { state, dispatch } = useContext(ComponentContext);
+  const { selectedComponent, errors = {}, hasValidated } = state;
+  const { page, toggleAddComponent = () => {} } = props;
+
+  const [isSaving, setIsSaving] = useState(false);
+  const hasErrors = hasValidationErrors(errors);
+
+  useEffect(() => {
+    dispatch({ type: Actions.SET_PAGE, payload: page.path });
+  }, [page.path]);
+
+  useLayoutEffect(() => {
+    if (hasValidated && !hasErrors) {
+      handleSubmit()
+        .then()
+        .catch((err) => {
+          console.error(err);
```

Comment:
Could be worth starting to call a simple client logger module that for now will just console.error out, but soon will make it simpler to call Sentry with errors from one place in the code?